consensus/src/rand/rand_gen/types.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_consensus_types::{
+    common::{Author, Round},
+    randomness::{RandMetadata, Randomness},
+};
+use std::collections::HashMap;
+
+#[derive(Clone, Debug)]
+pub(super) struct MockShare;
+
+#[derive(Clone, Debug)]
+pub(super) struct MockProof;
+
+impl Share for MockShare {
+    fn verify(
+        &self,
+        _rand_config: &RandConfig,
+        _rand_metadata: &RandMetadata,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+impl Proof for MockProof {
+    type Share = MockShare;
+
+    fn verify(
+        &self,
+        _rand_config: &RandConfig,
+        _rand_metadata: &RandMetadata,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    fn aggregate<'a>(
+        _shares: impl Iterator<Item = &'a RandShare<Self::Share>>,
+        _rand_config: &RandConfig,
+        rand_metadata: RandMetadata,
+    ) -> RandDecision<Self> {
+        RandDecision::new(Randomness::new(rand_metadata, vec![0u8; 32]), Self)
+    }
+}
+
+pub trait Share {
+    fn verify(&self, rand_config: &RandConfig, rand_metadata: &RandMetadata) -> anyhow::Result<()>;
+}
+
+pub trait Proof {
+    type Share: 'static;
+    fn verify(&self, rand_config: &RandConfig, rand_metadata: &RandMetadata) -> anyhow::Result<()>;
+
+    fn aggregate<'a>(
+        shares: impl Iterator<Item = &'a RandShare<Self::Share>>,
+        rand_config: &RandConfig,
+        rand_metadata: RandMetadata,
+    ) -> RandDecision<Self>
+    where
+        Self: Sized;
+}
+
+#[derive(Clone, Debug)]
+pub struct RandShare<S> {
+    author: Author,
+    metadata: RandMetadata,
+    share: S,
+}
+
+impl<S: Share> RandShare<S> {
+    pub fn new(author: Author, metadata: RandMetadata, share: S) -> Self {
+        Self {
+            author,
+            metadata,
+            share,
+        }
+    }
+
+    pub fn author(&self) -> &Author {
+        &self.author
+    }
+
+    pub fn metadata(&self) -> &RandMetadata {
+        &self.metadata
+    }
+
+    pub fn round(&self) -> Round {
+        self.metadata.round()
+    }
+
+    pub fn epoch(&self) -> u64 {
+        self.metadata.epoch()
+    }
+
+    pub fn verify(&self, rand_config: &RandConfig) -> anyhow::Result<()> {
+        self.share.verify(rand_config, &self.metadata)
+    }
+}
+
+pub struct RandDecision<P> {
+    randomness: Randomness,
+    proof: P,
+}
+
+impl<P: Proof> RandDecision<P> {
+    pub fn new(randomness: Randomness, proof: P) -> Self {
+        Self { randomness, proof }
+    }
+
+    pub fn verify(&self, rand_config: &RandConfig) -> anyhow::Result<()> {
+        self.proof.verify(rand_config, self.randomness.metadata())
+    }
+
+    pub fn randomness(&self) -> &Randomness {
+        &self.randomness
+    }
+
+    pub fn rand_metadata(&self) -> &RandMetadata {
+        self.randomness.metadata()
+    }
+}
+
+pub struct RandConfig {
```

Comment:
Will there be another `RandConfig` as part of NodeConfig or OnChainConfig that might confuse with this? If so, may be clarify the name properly?

Reply:
this is the on-chain version, we'll probably have a local version, but yes we should clarify names